src/main/scala/ir/Program.scala

```diff
@@ -179,17 +318,64 @@ class Procedure(
   }
 
 }
+class Parameter(var name: String, var size: Int, var value: Register) {
+  def toBoogie: BVariable = BParam(name, BitVecBType(size))
+  def toGamma: BVariable = BParam(s"Gamma_$name", BoolBType)
+}
+
+
+class Block private (var label: String,
+ var address: Option[Int],
+ val statements: IntrusiveList[Statement],
+ private var _jump: Jump,
+ val incomingJumps: mutable.HashSet[Block],
+) extends IntrusiveListElement, HasParent[Procedure] {
+  statements.foreach(_.setParent(this))
+  _jump.setParent(this)
+
+  statements.onInsert = x => x.setParent(this)
+  statements.onRemove = x => x.deParent()
+
+  def this(label: String, address: Option[Int], statements: IterableOnce[Statement], jump: Jump) = {
+    this(label, address, IntrusiveList.from(statements), jump, mutable.HashSet.empty)
+  }
+
+  def this(label: String, address: Option[Int], statements: IterableOnce[Statement]) = {
+    this(label, address, IntrusiveList.from(statements), GoTo(Seq(), Some(label + "_unknown")), mutable.HashSet.empty)
+  }
+
+  def this(label: String, address: Option[Int] = None) = {
+    this(label, address, IntrusiveList(), GoTo(Seq(), Some(label + "_unknown")), mutable.HashSet.empty)
+  }
+
+  def jump: Jump = _jump
+
+  def addGoToTargets(targets: mutable.Set[Block]): this.type = {
+    require(_jump.parent == this && _jump.isInstanceOf[GoTo])
+    _jump.asInstanceOf[GoTo].addAllTargets(targets)
+    this
+  }
+
+  def replaceJump(j: Jump): this.type = {
+    _jump.deParent()
+    j.setParent(this)
+    _jump = j
+    this
+  }
```

Comment:
Why do these return `this`? Is there a need for addGoToTargets to even exist?

Reply:
> Is there a need for addGoToTargets to even exist?

It's a transformation we might want but if needed it should be moved to GoTo not block

They return this for convenience of chaining modifications